File: leetcode/97.cpp
// 水题，通过递归实现 DP 就可以了。

class Solution {
public:
	map<pair<int, int>, bool> dp;

	bool isInterleave(const string &s1, int i1, const string &s2, int i2, const string &s3) {
		int n1 = s1.size(), n2 = s2.size();
		int n3 = s3.size();
		int i3 = i1 + i2;

		if (i3 == n3)
			return true;
		
		auto p = make_pair(i1, i2);
		auto it = dp.find(p);
		if (it != dp.end())
			return it->second;

		bool dpp = false;
		if (i1 < n1 && s1[i1] == s3[i3]) {
			dpp = dpp || isInterleave(s1, i1 + 1, s2, i2, s3);
		}
		if (i2 < n2 && s2[i2] == s3[i3]) {
			dpp = dpp || isInterleave(s1, i1, s2, i2 + 1, s3);
		}
		return dp[p] = dpp;
	}

	bool isInterleave(const string &s1, const string &s2, const string &s3) {
		int n1 = s1.size(), n2 = s2.size();
		int n3 = s3.size();
		if (n3 != n1 + n2)
			return false;

		dp.clear();
		return isInterleave(s1, 0, s2, 0, s3);
	}
};

File: leetcode/97.rs
impl Solution {
    fn wtf(s1: &[char], s2: &[char], s3: &[char]) -> bool {
        if s1.len() == 0 {
            return s2 == s3;
        } else if s2.len() == 0 {
            return s1 == s3;
        } else if s3.len() == 0 {
            return false;
        } else {
            let c1 = s1[0];
            let c3 = s3[0];
            if c1 == c3 && Self::wtf(&s1[1..], s2, &s3[1..]) {
                return true;
            }
            let c2 = s2[0];
            return c2 == c3 && Self::wtf(s1, &s2[1..], &s3[1..]);
        }
    }

    pub fn is_interleave(s1: String, s2: String, s3: String) -> bool {
        Self::wtf(
            &s1.chars().collect::<Vec<char>>(),
            &s2.chars().collect::<Vec<char>>(),
            &s3.chars().collect::<Vec<char>>(),
        )
    }
}
